rp2040-hal/src/uart/peripheral.rs

```diff
@@ -504,8 +504,8 @@ impl<D: UartDevice, P: ValidUartPinout<D>> embedded_io::ReadReady
 
 impl<D: UartDevice, P: ValidUartPinout<D>> embedded_io::Write for UartPeripheral<Enabled, D, P> {
     fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Error> {
-        self.write_full_blocking(buf);
-        Ok(buf.len())
+        let remaining = nb::block!(super::writer::write_raw(&self.device, buf)).unwrap(); // Infallible
```

Comment:
Might be worth repeating the embedded_io::Write rules here ("If the writer is not currently ready to accept more bytes (for example, its buffer is full), this function blocks until it is ready to accept least one byte.")

Reply:
and that write_raw only blocks if the FIFO is full and nothing is written to it.